### PR TITLE
Model for transfer of app ownership

### PR DESCRIPTION
**Pull request by janibonnevier**
#131

### CODE REVIEW

File: nextcloudappstore/core/models.py [hunk 1]
```diff
@@ -512,3 +512,62 @@ class Meta:
...
+
+        AppOwnershipTransfer.objects.create(app=app, new_owner=user)
+
+    An instance of AppOwnershipTransfer is deleted when the transfer it
+    represents is committed.
+    """
+
+    app = OneToOneField(
+        App, on_delete=CASCADE, related_name='ownership_transfer')
+    old_owner = ForeignKey(
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: would probably name this field **from**

File: nextcloudappstore/core/models.py [hunk 2]
```diff
@@ -512,3 +512,62 @@ class Meta:
...
+        User,
+        related_name='app_ownership_transfers_outgoing',
+        on_delete=CASCADE,
+        null=False)
+    new_owner = ForeignKey(
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: and this field **to**

File: nextcloudappstore/core/models.py [hunk 3]
```diff
@@ -512,3 +512,62 @@ class Meta:
...
+    following statement:
+
+        AppOwnershipTransfer.objects.create(app=app, new_owner=user)
+
+    An instance of AppOwnershipTransfer is deleted when the transfer it
+    represents is committed.
+    """
+
+    app = OneToOneField(
+        App, on_delete=CASCADE, related_name='ownership_transfer')
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: Models should be in wrapped in ', e.g. 'App' rather than App

File: nextcloudappstore/core/models.py [hunk 4]
```diff
@@ -512,3 +512,62 @@ class Meta:
...
+        User,
+        related_name='app_ownership_transfers_outgoing',
+        on_delete=CASCADE,
+        null=False)
+    new_owner = ForeignKey(
+        User,
+        related_name='app_ownership_transfers_incoming',
+        on_delete=CASCADE,
+        null=False)
+    is_accepted_old_owner = BooleanField(default=False)
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: Probably irrelevant to store if both users accepted it. If a request is created we will render it. If its approved we will delete the entry and change the owner on the app

Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: You do that further todwn btw

File: nextcloudappstore/core/models.py [hunk 5]
```diff
@@ -512,3 +512,62 @@ class Meta:
...
+
+    def save(self, force_insert=False, force_update=False, using=None,
+             update_fields=None):
+        if not self.id:  # a.k.a. "if object is being created"
+            self.old_owner = self.app.owner
+            if self.old_owner is self.new_owner:
+                raise RuntimeError(
+                    'Could not initiate transfer of app ownership. '
+                    + 'The proposed new owner already owns the app.')
+        return super().save()
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: arguments are not passed to super

File: nextcloudappstore/core/models.py [hunk 6]
```diff
@@ -512,3 +512,62 @@ class Meta:
...
+    represents is committed.
+    """
+
+    app = OneToOneField(
+        App, on_delete=CASCADE, related_name='ownership_transfer')
+    old_owner = ForeignKey(
+        User,
+        related_name='app_ownership_transfers_outgoing',
+        on_delete=CASCADE,
+        null=False)
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: https://docs.djangoproject.com/en/1.10/ref/models/fields/#null default is already False

File: nextcloudappstore/core/models.py (hunk 1 above)
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: and use settings.AUTH_USER_MODEL as user foreign key

Comment (janibonnevier) [inline nextcloudappstore/core/models.py]: `from` is a Python keyword. `from_user`?

File: nextcloudappstore/core/models.py (hunk 4 above)
Comment (janibonnevier) [inline nextcloudappstore/core/models.py]: I put it there for flexibility. But maybe we don't need that :)

File: nextcloudappstore/core/models.py [hunk 7]
```diff
@@ -512,3 +512,54 @@ class Meta:
...
+        'App', on_delete=CASCADE, related_name='ownership_transfer')
+    from_user = ForeignKey(
+        settings.AUTH_USER_MODEL,
+        related_name='app_ownership_transfers_outgoing',
+        on_delete=CASCADE)
+    to_user = ForeignKey(
+        settings.AUTH_USER_MODEL,
+        related_name='app_ownership_transfers_incoming',
+        on_delete=CASCADE)
+    is_accepted = BooleanField(default=False)
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: we can probably ditch that field

File: nextcloudappstore/core/models.py [hunk 8]
```diff
@@ -512,3 +512,54 @@ class Meta:
...
+        self.app.owner = self.to_user
+        self.app.save()
+        self.delete()
+
+    def save(self, *args, **kwargs):
+        if not self.id:  # a.k.a. "if object is being created"
+            self.from_user = self.app.owner
+            if self.from_user is self.to_user:
+                raise RuntimeError(
+                    'Could not initiate transfer of app ownership. '
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/models.py]: the + is not needed inside () afaik to concat strings

File: nextcloudappstore/core/tests/test_app_ownership_transfer.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+from django.contrib.auth.models import User
+from django.test import TestCase
+from nextcloudappstore.core.models import App, AppOwnershipTransfer
+
+
+class AppOwnershipTransferTest(TestCase):
+    def setUp(self):
+        self.user1 = User.objects.create_user(username='user1')
```
Comment (BernhardPosselt) [inline nextcloudappstore/core/tests/test_app_ownership_transfer.py]: AFAIK things run in setUp are not included in the transaction and you need to delete them again in the tearDown method. I'd probably move them to a separate method and call that one in each relevant block. Alternative is to delete them in a tearDown method

**Pull request merged**